File: packages/core/src/spring.test.ts
import { describe, it, expect } from 'vitest';
import { createSpring, estimateSpringDuration, SPRING_PRESETS } from './spring';

describe('createSpring', () => {
  it('starts at position 0', () => {
    const spring = createSpring();
    expect(spring.current().position).toBe(0);
    expect(spring.current().atRest).toBe(false);
  });

  it('converges to position 1', () => {
    const spring = createSpring({ stiffness: 200, damping: 20 });
    let state = spring.current();
    for (let i = 0; i < 300; i++) {
      state = spring.step(1 / 60);
    }
    expect(state.position).toBeCloseTo(1, 2);
    expect(state.atRest).toBe(true);
  });

  it('bouncy config overshoots past 1', () => {
    const spring = createSpring(SPRING_PRESETS.bouncy);
    let maxPos = 0;
    for (let i = 0; i < 300; i++) {
      const state = spring.step(1 / 60);
      if (state.position > maxPos) maxPos = state.position;
    }
    // bouncy should overshoot past 1
    expect(maxPos).toBeGreaterThan(1.01);
  });

  it('noWobble config does not overshoot significantly', () => {
    const spring = createSpring(SPRING_PRESETS.noWobble);
    let maxPos = 0;
    for (let i = 0; i < 300; i++) {
      const state = spring.step(1 / 60);
      if (state.position > maxPos) maxPos = state.position;
    }
    // noWobble should barely overshoot
    expect(maxPos).toBeLessThan(1.05);
  });

  it('reset returns to initial state', () => {
    const spring = createSpring();
    spring.step(1 / 60);
    spring.step(1 / 60);
    spring.reset();
    expect(spring.current().position).toBe(0);
  });
});

describe('estimateSpringDuration', () => {
  it('returns reasonable duration for default config', () => {
    const dur = estimateSpringDuration();
    expect(dur).toBeGreaterThan(200);
    expect(dur).toBeLessThan(5000);
  });

  it('snappy is faster than slow', () => {
    const snappy = estimateSpringDuration(SPRING_PRESETS.snappy);
    const slow = estimateSpringDuration(SPRING_PRESETS.slow);
    expect(snappy).toBeLessThan(slow);
  });

  it('all presets converge within 5 seconds', () => {
    for (const [name, config] of Object.entries(SPRING_PRESETS)) {
      const dur = estimateSpringDuration(config);
      expect(dur, `${name} should converge`).toBeLessThan(5000);
    }
  });
});

File: packages/android/src/test/kotlin/com/fadeanimation/SpringTest.kt
package com.fadeanimation

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * 验证 Android 弹簧求解器与 Web 端 @fade-animation/core 数值一致。
 * 断言镜像 core 的 spring.test.ts，保证四端弹簧手感统一。
 */
class SpringTest {

    @Test
    fun `starts at zero`() {
        val solver = SpringSolver()
        assertEquals(0f, solver.current().position, 0.0001f)
        assertFalse(solver.current().atRest)
    }

    @Test
    fun `converges to one`() {
        val solver = SpringSolver(SpringConfig(stiffness = 200f, damping = 20f))
        var state = solver.current()
        repeat(300) { state = solver.step(1f / 60f) }
        assertEquals(1f, state.position, 0.01f)
        assertTrue(state.atRest)
    }

    @Test
    fun `bouncy overshoots past one`() {
        val solver = SpringSolver(SpringPresets.BOUNCY)
        var maxPos = 0f
        repeat(300) {
            val state = solver.step(1f / 60f)
            if (state.position > maxPos) maxPos = state.position
        }
        assertTrue(maxPos > 1.01f, "bouncy should overshoot")
    }

    @Test
    fun `noWobble does not overshoot significantly`() {
        val solver = SpringSolver(SpringPresets.NO_WOBBLE)
        var maxPos = 0f
        repeat(300) {
            val state = solver.step(1f / 60f)
            if (state.position > maxPos) maxPos = state.position
        }
        assertTrue(maxPos < 1.05f, "noWobble should barely overshoot")
    }

    @Test
    fun `reset returns to initial`() {
        val solver = SpringSolver()
        solver.step(1f / 60f)
        solver.step(1f / 60f)
        solver.reset()
        assertEquals(0f, solver.current().position, 0.0001f)
    }

    @Test
    fun `estimate duration is reasonable`() {
        val dur = estimateSpringDuration()
        assertTrue(dur > 200)
        assertTrue(dur < 5000)
    }

    @Test
    fun `snappy is faster than slow`() {
        val snappy = estimateSpringDuration(SpringPresets.SNAPPY)
        val slow = estimateSpringDuration(SpringPresets.SLOW)
        assertTrue(snappy < slow)
    }

    @Test
    fun `all presets converge within 5 seconds`() {
        listOf(
            SpringPresets.GENTLE, SpringPresets.SNAPPY, SpringPresets.BOUNCY,
            SpringPresets.SLOW, SpringPresets.NO_WOBBLE
        ).forEach { config ->
            assertTrue(estimateSpringDuration(config) < 5000)
        }
    }
}
